Implement "tissue bridge ratio" metrics

valosekj opened the pull request:
## Checklist

<!-- Hi, and thank you for submitting a Pull Request! Please make sure to check the boxes below before submitting. -->

- [x] **PR Sidebar**: I've filled in each of the options within the PR sidebar. ([Reviewers](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#231-reviewers), [Assignees](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#232-assignees), [Labels](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#233-labels) (**exactly one** dark purple label!), [Milestone](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#234-milestone), [Development](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#235-development).)
- [ ] **Contributing Docs**: I've read the [Contributing guidelines](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)) to make sure my [branch](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing%3A-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#part-1-branches) and [pull request](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing%3A-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#part-2-pull-requests) meet SCT's guidelines. (Feel free to stop and fixup your commits before submitting.)
- [ ] **Tests**: I've added [relevant tests](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Testing) for my contribution, if necessary. I've also made sure that my contribution passes [all of the automated tests](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#24-checking-the-automated-tests) (which will be triggered after the PR is submitted).
- [ ] **Documentation**: I've updated the [relevant documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages.

## Description

This PR adds two new metrics (and one old one) to the `sct_analyze_lesion` output:
```
- interpolated_total_bridge_width `[mm]`: total width of spared tissue in the **interpolated midsagittal slice** (i.e. dorsal + ventral)
- ventral_bridge_ratio: ventral midsagittal tissue bridge width divided by total midsagittal tissue bridge width
- dorsal_bridge_ratio: dorsal midsagittal tissue bridge width divided by total midsagittal tissue bridge width
```


<details>
<summary>example outputs on different lesions</summary>

<img width="200" alt="Screenshot 2025-06-14 at 5 48 47 PM" src="https://github.com/user-attachments/assets/93254654-44fc-4ec8-be33-6091f2c02d92" />

```
...
  Midsagittal dorsal tissue bridge width: 0.77 mm
  Midsagittal ventral tissue bridge width: 0.76 mm
  Midsagittal dorsal bridge ratio: 50.14 %
  Midsagittal ventral bridge ratio: 49.86 %
...
```

<img width="200" alt="Screenshot 2025-06-14 at 5 48 51 PM" src="https://github.com/user-attachments/assets/315d0470-6627-4c54-843c-29f3de0eca59" />

```
...
  Midsagittal dorsal tissue bridge width: 0.74 mm
  Midsagittal ventral tissue bridge width: 2.95 mm
  Midsagittal dorsal bridge ratio: 20.02 %
  Midsagittal ventral bridge ratio: 79.98 %
...
```

<img width="200" alt="Screenshot 2025-06-14 at 5 48 53 PM" src="https://github.com/user-attachments/assets/acdb0cd1-e567-435e-871f-6baeabdc08c5" />

```
...
  Midsagittal dorsal tissue bridge width: 0.0 mm
  Midsagittal ventral tissue bridge width: 4.34 mm
  Midsagittal dorsal bridge ratio: 0.0 %
  Midsagittal ventral bridge ratio: 100.0 %
...
```

</details>

> [!NOTE]
> Right now, I have implemented the tissue bridge ratio metrics only for the midsagittal tissue bridges. However, since `sct_analyze_lesion` computes tissue bridges for each sagittal slice (midsagittal and parasagittal), I can easily extend the ratio computation to include the parasagittal slices as well.

Additionally, the PR: 
 - adds a [new test](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/442cdc3a2762c68cfdd65003147439d1d000a526/testing/cli/test_cli_sct_analyze_lesion.py#L322-L361) to test that the script exits when no lesion is found in the input image
 - fixes bugs

## How to test the PR?

Download the attached zip archive with sample data and then run:

```bash
cd $SCT_DIR
git fetch
git checkout jv/4928-implement_tissue_bridge_ratio

sct_analyze_lesion -m t2_lesion.nii.gz -s t2_seg.nii.gz
sct_analyze_lesion -m t2_lesion2.nii.gz -s t2_seg.nii.gz
sct_analyze_lesion -m t2_lesion3.nii.gz -s t2_seg.nii.gz
```

[4933-sample_data_tissue_bridge_ratios.zip](https://github.com/user-attachments/files/20738570/4933-sample_data_tissue_bridge_ratios.zip)

## Linked issues

Fixes https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4928

valosekj commented:
Hey @dpfyffer! I've implemented the first version of tissue bridge ratio metrics. Please, let me know what you think 😊

dpfyffer commented:
@valosekj awesome, thank you!
Will test and let you know.

valosekj commented:
Tested on @dpfyffer's machine, and it works! 

@joshuacwnewton, the pull request is now ready for review 🙂

joshuacwnewton commented on spinalcordtoolbox/scripts/sct_analyze_lesion.py:
```diff
@@ -333,6 +335,32 @@ def _measure_axial_damage_ratio(self, im_data, p_lst, idx):
         printv('  Maximum axial damage ratio: ' + str(np.round(maximum_axial_damage_ratio, 2)),
                self.verbose, type='info')
 
+    def _compute_issue_bridge_ratio(self, idx):
```
nitpick, but: typo!

```suggestion
    def _compute_tissue_bridge_ratio(self, idx):
```

(will have to fix the caller site as well)

joshuacwnewton commented on spinalcordtoolbox/scripts/sct_analyze_lesion.py:
```diff
@@ -333,6 +335,32 @@ def _measure_axial_damage_ratio(self, im_data, p_lst, idx):
+        ventral_bridge_ratio = ventral_bridge_width / total_bridge_width * 100 if total_bridge_width > 0 else 0
+
+        # Save the ratios to the measure DataFrame
+        self.measure_pd.loc[idx, 'dorsal_bridge_ratio [%]'] = dorsal_bridge_ratio
+        self.measure_pd.loc[idx, 'ventral_bridge_ratio [%]'] = ventral_bridge_ratio
```
This should be something we can test fairly easily, right? Since we're already checking both bridge widths in the "compute_expected_measurements" function, the calculation of the "expected" ratio seems very straightforward too. :D

valosekj replied:
That's an excellent point! I actually started writing a test for these new ratio metrics, but then I realized that we are actually not computing (and testing) `interpolated_dorsal_bridge_width` and `interpolated_ventral_bridge_width` in our test suite. These interpolated bridges are used to compute `dorsal_bridge_ratio` and `ventral_bridge_ratio`:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/0f97779b53b3f81d6698b7382920f9914fba290c/spinalcordtoolbox/scripts/sct_analyze_lesion.py#L345-L347 

In other words, if we want to test `dorsal_bridge_ratio` and `ventral_bridge_ratio`, we need to add computation of `interpolated_dorsal_bridge_width` and `interpolated_ventral_bridge_width`.

valosekj replied:
> ```python
>             # We don't actually store the total interpolated bridge width in the final output, so delete it
>             del tissue_bridges['interpolated_total_bridge_width [mm]']
> ```

That's a good point, actually! Total interpolated bridge width is also useful, so why not store it? I added it to the main script in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4933/commits/69dc10d72000c1be4ea12a87371814ad5245b82f.

valosekj replied:
> It's okay! I spent a little time implementing the test, and it seems to work OK?

This is awesome, @joshuacwnewton! Thank you a lot! I took the liberty of including your code in my recent commits.

valosekj replied:
> The values are the same, but one value has slightly more precision than the other, so it is greater than expected. I'm not quite sure the best fix for this? Maybe rounding both values to the same number of decimal points before comparing? 🤔

Good point! I tried rounding to 10 decimal places, and it works, so I implemented it in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4933/commits/f28111b5a40ac1ac8335815fad324dcb8ff11f02

joshuacwnewton commented on spinalcordtoolbox/scripts/sct_analyze_lesion.py:
```diff
@@ -1027,6 +1063,10 @@ def label_lesion(self):
         self.measure_pd['label'] = [label for label in np.unique(im_2save.data) if label]
         printv('Lesion count = ' + str(len(self.measure_pd['label'])), self.verbose, 'info')
 
+        # Exit the script if no lesion is found
+        if len(self.measure_pd['label']) == 0:
+            printv('ERROR: No lesion found in the input image.', self.verbose, 'error')
```
Maybe we could add a test for this as well? I'm curious what happen if an empty lesion file is passed to `sct_analyze_lesion`. 

Does it fail earlier? Or was it not failing at all, hence the need for the check?

valosekj replied:
Yes, in the case of an empty lesion, the script is terminated by exactly this line

```shell
Lesion count = 0
ERROR: No lesion found in the input image.
``` 

Speaking about the script termination, as you've opened this question, I've been actually thinking for a while about changing this behaviour slightly. Because currently, if there's no lesion, the script ends with an error call. If `sct_analyze_lesion` is called from a bash script run by our `sct_run_batch` wrapper, this then, in turn, results in termination of the entire bash script. I've been using the following workaround to not terminate the bash script in the case of an empty lesion.

```
  # Check if the lesion is empty, if so, skip sct_analyze_lesion (because otherwise sct_analyze_lesion would stop the script)
  status=$($SCT_DIR/python/envs/venv_sct/bin/python -c "import nibabel as nib; import numpy as np; print('empty' if np.all(nib.load('${file}_lesion.nii.gz').get_fdata() == 0) else 'not empty')")
  if [ "$status" = "empty" ]; then
      echo "❌ [$(date '+%Y-%m-%d %H:%M:%S')] ${file}: No lesion found in the input image. Skipping lesion analysis." >> "${PATH_LOG}/T2w_lesion_analysis.log"
  else
      sct_analyze_lesion -m ${file}_lesion.nii.gz -s ${file}_seg.nii.gz -f label_T2w -qc ${PATH_QC} -qc-subject ${file}
      echo "✅ [$(date '+%Y-%m-%d %H:%M:%S')] ${file}: Lesion analysis completed successfully." >> "${PATH_LOG}/T2w_lesion_analysis.log"
  fi
```

But I think it should be pretty easy by doing something like this:

```python
# Exit the script gracefully if no lesion is found
if len(self.measure_pd['label']) == 0:
    printv('WARNING: No lesion found in the input image. Exiting.', self.verbose, 'warning')
    print('Analysis completed with no results (no lesion found).')
    sys.exit(0)  # Exit with success code
```

valosekj replied:
I drafted a possible workaround and a new test in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4933/commits/b51ae77e33004907272394c35c11081262d28e68. Let me know what you think :-)

valosekj replied:
After a discussion during the 2025-07-31 SCT dev meeting, I switched back to exit code 1 in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4933/commits/13b6283e1cf3f8f42d383e59070fc5e3060e7ced

joshuacwnewton commented on spinalcordtoolbox/scripts/sct_analyze_lesion.py:
```diff
@@ -333,6 +335,32 @@ def _measure_axial_damage_ratio(self, im_data, p_lst, idx):
+        # Save the ratios to the measure DataFrame
+        self.measure_pd.loc[idx, 'dorsal_bridge_ratio [%]'] = dorsal_bridge_ratio
+        self.measure_pd.loc[idx, 'ventral_bridge_ratio [%]'] = ventral_bridge_ratio
+        printv(f'  Midsagittal dorsal bridge ratio: {np.round(dorsal_bridge_ratio, 2)} %',
+               self.verbose, type='info')
+        printv(f'  Midsagittal ventral bridge ratio: {np.round(ventral_bridge_ratio, 2)} %',
+               self.verbose, type='info')
```
Tiny nit, but you can use f-string formatting to get the desired output:

```suggestion
        printv(f'  Midsagittal dorsal bridge ratio: {dorsal_bridge_ratio:.2f} %',
               self.verbose, type='info')
        printv(f'  Midsagittal ventral bridge ratio: {ventral_bridge_ratio:.2f} %',
               self.verbose, type='info')
```

joshuacwnewton commented on spinalcordtoolbox/scripts/sct_analyze_lesion.py:
```diff
@@ -333,6 +335,32 @@ def _measure_axial_damage_ratio(self, im_data, p_lst, idx):
+        total_bridge_width = dorsal_bridge_width + ventral_bridge_width
+
+        # Compute the dorsal and ventral bridge ratios
+        dorsal_bridge_ratio = dorsal_bridge_width / total_bridge_width * 100 if total_bridge_width > 0 else 0
+        ventral_bridge_ratio = ventral_bridge_width / total_bridge_width * 100 if total_bridge_width > 0 else 0
```
nit: I think maybe we want `0.0` instead of `0` just to keep the float type consistent?

Also, these one-liners are getting a little long, so it might help with readability to split them up:

```suggestion
    if total_bridge_width > 0:
        dorsal_bridge_ratio = (dorsal_bridge_width / total_bridge_width) * 100
        ventral_bridge_ratio = (ventral_bridge_width / total_bridge_width) * 100
    else:
        dorsal_bridge_ratio = 0.0
        ventral_bridge_ratio = 0.0
```

valosekj replied:
sounds good! accepting the suggestion

joshuacwnewton reviewed (changes requested):
This mostly looks good to me! The calculation is very straightforward, so my comments are mostly nitpicky re: code readability and testing.

valosekj commented:
To test for the case described in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4933#issuecomment-3150529290, I added a new dummy lesion file with two lesions along the spinal cord RL axis in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4933/commits/6f16882913b045e1efefc5e087e4d592cabe15a9:

<img width="1000" height="1056" alt="image" src="https://github.com/user-attachments/assets/92f7f6d2-4e63-44ff-b5e9-e049df814231" />

When running our existing test suite on this new dummy lesion, I discovered an indentation bug in the dictionary assignment 
when computing expected measurements for multiple lesions. I fixed it in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4933/commits/442cdc3a2762c68cfdd65003147439d1d000a526

joshuacwnewton commented on testing/cli/test_cli_sct_analyze_lesion.py:
```diff
@@ -143,8 +158,7 @@ def compute_expected_measurements(lesion_params, path_seg=None):
             **tissue_bridges
         }
-
-    expected_measurements_dict[idx] = measurements
```
Nice catch! :hearts:

joshuacwnewton reviewed (approved):
LGTM! Thanks so much for this contribution. :hearts:
